### ISSUE TITLE
[LoggerMessageAttribute] generates CS0757 and CS0102 when combined with XAML and certain NuGet package references.

### ISSUE THREAD

**Issue by daviddunson** [dotnet/wpf#9589]
### Description

Build will fail when targeting specific NuGet packages in addition to compiling with a .XAML file. If the .XAML file is removed, or the NuGet package is removed, the build will succeed. It is when .XAML, [LoggerMessage] and specific packages are referenced.

This was discovered while embedding an ASP.NET Core host inside a WPF application. These packages were referenced in a class library that was referenced by the WPF project:

```
    <PackageReference Include="IdentityServer4" Version="4.1.2" />
    <PackageReference Include="Microsoft.AspNetCore.Authentication.JwtBearer" Version="8.0.8" />
    <PackageReference Include="Microsoft.AspNetCore.Mvc.NewtonsoftJson" Version="8.0.8" />
    <PackageReference Include="Swashbuckle.AspNetCore" Version="6.7.0" />
    <PackageReference Include="System.Linq.Async" Version="6.0.1" />
    <PackageReference Include="System.Text.Encodings.Web" Version="8.0.0" />
 ```

I was able to reproduce the error by creating a new WPF project and adding a direct reference to one of these packages.

### Reproduction Steps

Create a new C# WPF Application targeting .NET 8.

Add the following package references:

```
  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.Authentication.JwtBearer" Version="8.0.8" />
    <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="8.0.1" />
  </ItemGroup>
```

Add a class containing a [LoggerMessage] reference:

```
    public static partial class LoggerExtensions
    {
        [LoggerMessage(LogLevel.Error, "Your message: {Message}")]
        public static partial void SomethingFailed(this ILogger logger, string message);
    }
```

Try to build the project.

### Expected behavior

Build succeeded.

### Actual behavior

Build failed.

```
1>...obj\Debug\net8.0-windows\Microsoft.Extensions.Logging.Generators\Microsoft.Extensions.Logging.Generators.LoggerMessageGenerator\LoggerMessage.g.cs(13,36,13,51): error CS0757: A partial method may not have multiple implementing declarations
1>...obj\Debug\net8.0-windows\Microsoft.Extensions.Logging.Generators\Microsoft.Extensions.Logging.Generators.LoggerMessageGenerator\LoggerMessage.g.cs(9,151,9,176): error CS0102: The type 'LoggerExtensions' already contains a definition for '__SomethingFailedCallback'
```

### Regression?

Project will build after rolling back to Visual Studio 17.9.

### Known Workarounds

Move the members with [LoggerMessageAttribute] to a class library that does not contain a .XAML file and a reference to one of the packages that trigger the failure.

### Configuration

Windows Version 11 23H2 (x64)
Visual Studio 17.11 (or 17.10)
.NET SDK 8.0.400

**Comment by tarekgh** [comment]
Could be similar to https://github.com/dotnet/extensions/pull/5024 and the fix should be in wpf instead? 

CC @xakep139 @joperezr

**Comment by ericstj** [comment]
I see this is in Future milestone - you might want to have a look sooner than that.  Seems like something changed here recently that's regressing folks.  @singhashish-wpf

**Comment by NicoKno** [comment]
We have this issue currently, and cannot get rid of it. Are there any workarounds we can try?

**Comment by trovialdo** [comment]
@NicoKno  #7624 contains a work around, but the work around itself has downsides as well (at least for us).

This issue is quite an annoying for us and are eager to see resolved. Thanks franchyd for the list.

What is the main issue through which this is tracked and will be resolved so we can only monitor that one to know when it is resolved?

**Comment by franchyd** [comment]
Still no updates?

**Comment by daviddunson** [comment]
WORK AROUND: Move the .XAML files to a separate assembly, or move the classes that use [LoggerMessage] to a separate assembly.

Following the MVVM pattern, if you have XAML files writing to the log files, this code is probably a good candidate for a view model or even a model or service which should handle logging. If you have code behind in your view that is performing low level code such as native Win32 API calls, you can call methods on your view model to perform logging.

**Comment by franchyd** [comment]
> WORK AROUND: Move the .XAML files to a separate assembly, or move the classes that use [LoggerMessage] to a separate assembly.
> 
> Following the MVVM pattern, if you have XAML files writing to the log files, this code is probably a good candidate for a view model or even a model or service which should handle logging. If you have code behind in your view that is performing low level code such as native Win32 API calls, you can call methods on your view model to perform logging.

This issue affects all “built in” source generators not just logger message.   That list is:

- System.Text.Json.SourceGeneration.JsonSourceGenerator
- Microsoft.AspNetCore.Razor.SourceGenerators.RazorSourceGenerator
- Microsoft.AspNetCore.OpenApi.OpenApiDocumentGenerator
- Microsoft.AspNetCore.Mvc.Generator
- Microsoft.Extensions.Logging.Generators.LoggerMessageGenerator


the one that impacts me is the JSON one.

**Comment by ericstj** [comment]
TL;DR A simple workaround here would be to make sure conflict resolution runs in WPF's generated project:
```xml
  <Target Name="_EnsureConflictResolution"
          Condition="'$(_TargetAssemblyProjectName)' != ''"
          AfterTargets="ResolveProjectReferences"
          DependsOnTargets="_HandlePackageFileConflicts" />
```
Simply add this to any WPF project with this issue.

----------------------------------

Bumping this back to 10.0.0 milestone.  @singhashish-wpf @pchaurasia14 please prioritize.

Just to reiterate what I said before -- this is happening because WPF's targets run package resolution but don't resolve any conflicts that might happen between those packages and framework assemblies/analyzers (or other packages).  Conflict resolution has been part of .NET Core since 2.0 and it's what enabled the framework to transition from entirely package-based (with 100+ packages) to using Targeting Packs.  It's also what continues to permit newer overlapping packages - like System.Text.Json, or Microsoft.Extensions.Logging.Abstractions - so that folks can use a package newer than their framework and still have that work.

This is broken because of this: https://github.com/dotnet/wpf/blob/20ed0120dae50dee312cddb5d9239821be72535b/src/Microsoft.DotNet.Wpf/src/PresentationBuildTasks/Microsoft.WinFX.targets#L346-L349

So WPF is using the `ResolveProjectReferences` target to try and trigger *package* resolution 😕 Seems a dubious choice.

As I mentioned before they end up raising a lot more than just source generators (eg: `ResolvedCompileFileDefinitions`) - it just so happens that they don't run the other targets that would end up copying these over to `ReferencePath` items that would reach the compiler.

See above for a better workaround.  This does use a couple _private_ identifiers: `_TargetAssemblyProjectName` to determine if it's in WPF's generated project, and `_HandlePackageFileConflicts`.

A better fix in WPF might be to change what targets are run to resolve packages.  I don't know why ResolveProjectReferences was chosen here: https://github.com/dotnet/wpf/commit/bd67c8f87342531d26bad5b4a84f3420e5290ab7#diff-681ca681001fbc20488df7fdbb46a8c40ac2df9dc2210df5ad869aee6a56df16R393-R394

Instead I think would suggest using  `$(ResolveAssemblyReferencesDependsOn)` since this [seems to be](https://github.com/search?q=repo%3Adotnet%2Fsdk%20ResolveAssemblyReferencesDependsOn&type=code) how the SDK defines the set of targets it needs to raise packages and resolve their conflicts.  @dsplaisted what do you think?

**Comment by dsplaisted** [comment]
I don't know a lot about how the WPF generated projects work.  The suggestion to switch to using `ResolveAssemblyReferencesDependsOn` seems to make sense, but I think the temp project build may be somewhat fragile so it might be easy to change.
